document-portal/document-portal-fuse.c

```diff
@@ -3155,44 +3265,57 @@ xdp_fuse_init (GError **error)
       return FALSE;
     }
 
-  main_ch = fuse_mount (path, &args);
-  if (main_ch == NULL)
+  if (fuse_parse_cmdline (&args, &opts) != 0)
     {
       fuse_opt_free_args (&args);
-      g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_FAILED,
-                   "Can't mount fuse fs on %s: %s", path, g_strerror (errno));
+      g_set_error (error, XDG_DESKTOP_PORTAL_ERROR,
+                   XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Impossible to parse command line");
       return FALSE;
     }
 
-  session = fuse_lowlevel_new (&args, &xdp_fuse_oper,
-                               sizeof (xdp_fuse_oper), NULL);
-  if (session == NULL)
+  g_mutex_init (&thread_data.lock);
+  g_cond_init (&thread_data.cond);
+
+  thread_data.args = &args;
+  thread_data.loop_config.clone_fd = opts.clone_fd;
+  thread_data.loop_config.max_idle_threads = opts.max_idle_threads;
+
+  g_mutex_lock (&thread_data.lock);
+  XDP_AUTOLOCK (session);
+  fuse_thread = g_thread_new ("fuse mainloop", xdp_fuse_thread, &thread_data);
+
+  while (session == NULL && thread_data.error == NULL)
+    g_cond_wait (&thread_data.cond, &thread_data.lock);
+
+  fuse_opt_free_args (&args);
+
+  g_mutex_unlock (&thread_data.lock);
+  g_cond_clear (&thread_data.cond);
+  g_mutex_clear (&thread_data.lock);
+
+  if (thread_data.error != NULL)
     {
-      fuse_opt_free_args (&args);
-      g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_FAILED,
-                   "Can't create fuse session");
+      g_propagate_error (error, g_steal_pointer (&thread_data.error));
       return FALSE;
     }
-  fuse_session_add_chan (session, main_ch);
 
-  fuse_thread = g_thread_new ("fuse mainloop", xdp_fuse_mainloop, session);
-
-  fuse_opt_free_args (&args);
+  g_assert (session != NULL);
 
   return TRUE;
 }
 
 void
 xdp_fuse_exit (void)
 {
-  if (!destroyed && session)
-    fuse_session_exit (session);
+  XDP_AUTOLOCK (session);
```

Comment:
The code in xdp_fuse_thread tries to reacquire this lock after the main loop exits. Are we actually going to be able to join the thread without first releasing this lock?

Reply:
So, well this can be seen as an early join from my POV (also reason why there was not a thread join before) because:
 - When the fuse loop is over (in case of normal exit), we destroy it
   - That leads to `on_fuse_unmount` be called and thus to stop the main loop
 -  xdp_fuse_exit is then called, and it waits for the session to be completely nullified and the thread to be over via this lock

So per se this function would be just uneeded in this case.
However, the reason for it to be here was to handle the case in which something else stopped the main loop (and not the fuse destruction), and in that case indeed we might end up in a dead lock when trying to join the thread.

As per this I've reduced the scope of the lock. In such case however we should not unset the session on exit (iirc it was suggested before, but given that `fuse_session_exit` only makes the fuse loop to exit we should definitely not clear the session at that level). Instead we need to ensure that the fuse thread will nullify it.